From: Dave Airlie <airlied@redhat.com>
Date: Tue Jan 21 01:47:46 2014 -0500
Subject: drm/mgag200: on cards with < 2MB VRAM default to 16-bit
Patch-mainline: Upstream
Git-commit: 918be888d613e58938338b4b0c895de97579173d

References: bnc#882324
Signed-off-by: Egbert Eich <eich@suse.com>

This aligns with what the userspace -mga driver does in
the same situation.

Signed-off-by: Dave Airlie <airlied@redhat.com>
---
 drivers/gpu/drm/mgag200/mgag200_fb.c   |    7 ++++++-
 drivers/gpu/drm/mgag200/mgag200_main.c |    5 ++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/mgag200/mgag200_fb.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/mgag200/mgag200_fb.c
@@ -275,6 +275,11 @@ int mgag200_fbdev_init(struct mga_device
 {
 	struct mga_fbdev *mfbdev;
 	int ret;
+	int bpp_sel = 32;
+
+	/* prefer 16bpp on low end gpus with limited VRAM */
+	if (IS_G200_SE(mdev) && mdev->mc.vram_size < (2048*1024))
+		bpp_sel = 16;
 
 	mfbdev = kzalloc(sizeof(struct mga_fbdev), GFP_KERNEL);
 	if (!mfbdev)
@@ -290,7 +295,7 @@ int mgag200_fbdev_init(struct mga_device
 		return ret;
 	}
 	drm_fb_helper_single_add_all_connectors(&mfbdev->helper);
-	drm_fb_helper_initial_config(&mfbdev->helper, 32);
+	drm_fb_helper_initial_config(&mfbdev->helper, bpp_sel);
 
 	return 0;
 }
--- linux-3.0-SLE11-SP3.orig/drivers/gpu/drm/mgag200/mgag200_main.c
+++ linux-3.0-SLE11-SP3/drivers/gpu/drm/mgag200/mgag200_main.c
@@ -242,7 +242,10 @@ int mgag200_driver_load(struct drm_devic
 	dev->mode_config.funcs = (void *)&mga_mode_funcs;
 	dev->mode_config.min_width = 0;
 	dev->mode_config.min_height = 0;
-	dev->mode_config.preferred_depth = 24;
+	if (IS_G200_SE(mdev) && mdev->mc.vram_size < (2048*1024))
+		dev->mode_config.preferred_depth = 16;
+	else
+		dev->mode_config.preferred_depth = 24;
 	dev->mode_config.prefer_shadow = 1;
 
 	r = mgag200_modeset_init(mdev);
